### PR TITLE
fix: side effects detect of destructing array

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
* **#2853** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2853?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2853?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs [hunk 1]
```diff
@@ -364,13 +375,28 @@ impl<'a> SideEffectDetector<'a> {
...
           }
         }
-        let is_destructuring = matches!(
-          declarator.id.kind,
-          BindingPatternKind::ArrayPattern(_) | BindingPatternKind::ObjectPattern(_)
-        );
-
-        is_destructuring
-          || declarator.init.as_ref().is_some_and(|init| self.detect_side_effect_of_expr(init))
+        match &declarator.id.kind {
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs:378]: https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/js_ast/js_ast_helpers.go#L2214-L2234

File: crates/rolldown/tests/esbuild/default/comment_preservation/artifacts.snap [hunk 2]
```diff
@@ -203,17 +203,13 @@ console.log(require.resolve(
...
-let [] = [];
 let [ ...s] = [...s];
 let [ ...s2] = [...s2];
 let {} = {};
 let {} = {};
 let { ...s3 } = { ...s3 };
 let { ...s4 } = { ...s4 };
 let [x] = [x];
-let [x2] = [x2];
-let [x3] = [x3];
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/default/comment_preservation/artifacts.snap:216]: https://hyrious.me/esbuild-repl/?version=0.23.0&b=e%00entry.ts%00let+%5B%0A++%2F*+foo+*%2F%0A%5D+%3D+%5B%0A++%2F*+bar+*%2F%0A%5D%3B%0Alet+%5B%0A++%2F%2F+foo%0A%5D+%3D+%5B%0A++%2F%2F+bar%0A%5D%3B%0A%0Alet+%5Bx2%5D+%3D+%5Bx2%5D%3B%0Alet+%5Bx3%5D+%3D+%5Bx3%5D%3B%0A&b=%00file.js%00&o=%7B%0A++treeShaking%3A+true%2C%0A++external%3A+%5B%22c%22%2C+%22a%22%2C+%22b%22%5D%2C%0A%22bundle%22%3A+true%2C%0Aformat%3A+%22esm%22%0A%7D, should has same output with esbuild in `bundle` mode

**Review by hyf0 — APPROVED** [review]
Nice

**Pull request merged**